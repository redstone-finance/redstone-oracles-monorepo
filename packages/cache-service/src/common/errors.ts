export const EMPTY_DATA_PACKAGE_RESPONSE_ERROR_CODE = 513;
