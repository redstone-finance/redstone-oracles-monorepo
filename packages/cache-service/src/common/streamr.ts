import * as pako from "pako";
import { StreamrClient } from "streamr-client";
import config from "../config";

export { StreamPermission, StreamrClient, Subscription } from "streamr-client";

export const getStreamIdForNodeByEvmAddress = (evmAddress: string) =>
  config.streamrStreamNamePattern.replaceAll("{evmAddress}", evmAddress);

export const doesStreamExist = async (
  streamr: StreamrClient,
  streamId: string
): Promise<boolean> => {
  try {
    await streamr.getStream(streamId);
    return true;
  } catch (error) {
    return false;
  }
};

export const compressMsg = (data: unknown) => {
  const dataStringified = JSON.stringify(data);
  return pako.deflate(dataStringified);
};

export const decompressMsg = <T = unknown>(msg: Uint8Array): T => {
  const stringifiedData = pako.inflate(msg, {
    to: "string",
  });
  return JSON.parse(stringifiedData) as T;
};
