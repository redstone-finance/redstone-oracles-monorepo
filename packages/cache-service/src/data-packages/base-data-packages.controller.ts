import {
  Body,
  Get,
  Header,
  HttpException,
  HttpStatus,
  Param,
  Post,
  Query,
  ServiceUnavailableException,
  UsePipes,
  ValidationPipe,
} from "@nestjs/common";
import config from "../config";
import {
  BulkPostRequestBody,
  DataPackagesResponse,
  DataPackagesStatsResponse,
  GetDataPackagesStatsQuery,
} from "./data-packages.interface";
import { DataPackagesService } from "./data-packages.service";

@UsePipes(new ValidationPipe({ transform: true }))
export abstract class BaseDataPackagesController {
  constructor(protected readonly dataPackagesService: DataPackagesService) {}

  protected abstract readonly allowExternalSigners: boolean;

  protected static async validateDataServiceId(dataServiceId: string) {
    const isDataServiceIdValid =
      await DataPackagesService.isDataServiceIdValid(dataServiceId);
    if (!isDataServiceIdValid) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          error: "Data service id is invalid",
        },
        HttpStatus.BAD_REQUEST
      );
    }
  }

  @Get("latest/:DATA_SERVICE_ID")
  @Header("Cache-Control", "max-age=5")
  async getAllLatest(
    @Param("DATA_SERVICE_ID") dataServiceId: string
  ): Promise<DataPackagesResponse> {
    await BaseDataPackagesController.validateDataServiceId(dataServiceId);
    return await this.dataPackagesService.getLatestDataPackagesWithSameTimestampWithCache(
      dataServiceId,
      false,
      this.allowExternalSigners
    );
  }

  @Get("latest-not-aligned-by-time/:DATA_SERVICE_ID")
  @Header("Cache-Control", "max-age=5")
  async getMostRecent(
    @Param("DATA_SERVICE_ID") dataServiceId: string
  ): Promise<DataPackagesResponse> {
    await BaseDataPackagesController.validateDataServiceId(dataServiceId);
    return await this.dataPackagesService.getLatestDataPackagesWithCache(
      dataServiceId,
      false,
      this.allowExternalSigners
    );
  }

  @Get("historical/:DATA_SERVICE_ID/:TIMESTAMP")
  @Header("Cache-Control", "max-age=5")
  async getDataPackagesByTimestamp(
    @Param("DATA_SERVICE_ID") dataServiceId: string,
    @Param("TIMESTAMP") timestamp: string
  ): Promise<DataPackagesResponse> {
    if (!config.enableHistoricalDataServing) {
      throw new ServiceUnavailableException(
        `historical/* routes are not enabled in this cache-service configuration`
      );
    }
    await BaseDataPackagesController.validateDataServiceId(dataServiceId);

    return await DataPackagesService.getDataPackagesByTimestamp(
      dataServiceId,
      Number(timestamp),
      false,
      this.allowExternalSigners
    );
  }

  @Get("stats/:PERIOD")
  // eslint-disable-next-line @typescript-eslint/class-methods-use-this
  async getStats(
    @Query() query: GetDataPackagesStatsQuery,
    @Param("PERIOD") period: string
  ): Promise<DataPackagesStatsResponse> {
    if (query["api-key"] !== config.apiKeyForAccessToAdminRoutes) {
      throw new HttpException(
        {
          status: HttpStatus.UNAUTHORIZED,
          error: "Incorrect api-key query param",
        },
        HttpStatus.UNAUTHORIZED
      );
    }

    const now = Date.now();
    return await DataPackagesService.getDataPackagesStats({
      fromTimestamp: now - Number(period),
      toTimestamp: now,
    });
  }

  @Post("bulk")
  async addBulk(@Body() body: BulkPostRequestBody) {
    if (!config.enableDirectPostingRoutes) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          error: "Data package posting routes are disabled",
        },
        HttpStatus.BAD_REQUEST
      );
    }

    const signerAddress = DataPackagesService.verifyRequester(body);

    const dataPackagesToSave =
      await DataPackagesService.prepareReceivedDataPackagesForBulkSaving(
        body.dataPackages,
        signerAddress
      );

    this.dataPackagesService
      .broadcast(dataPackagesToSave, signerAddress)
      .catch(() => {}); // ignore errors as they are logged by the service
  }
}
