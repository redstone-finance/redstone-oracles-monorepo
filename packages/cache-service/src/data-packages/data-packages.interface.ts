import { SignedDataPackagePlainObj } from "redstone-protocol";

export interface ReceivedDataPackage extends SignedDataPackagePlainObj {
  sources?: object;
}
