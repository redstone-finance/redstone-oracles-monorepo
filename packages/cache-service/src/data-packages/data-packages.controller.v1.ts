import { Controller } from "@nestjs/common";
import { BaseDataPackagesController } from "./base-data-packages.controller";

@Controller("data-packages")
export class DataPackagesControllerV1 extends BaseDataPackagesController {
  protected readonly allowExternalSigners = false;
}
