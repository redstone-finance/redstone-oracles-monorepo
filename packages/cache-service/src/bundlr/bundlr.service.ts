import { Injectable, Logger } from "@nestjs/common";
import { CachedDataPackage } from "../data-packages/data-packages.model";
import Bundlr from "@bundlr-network/client";
import config from "../config";

const REDSTONE_TYPE_TAG_VALUE = "redstone-oracles";

@Injectable()
export class BundlrService {
  private readonly logger = new Logger(BundlrService.name);
  private bundlrClient: Bundlr;

  constructor() {
    if (config.enableArchivingOnArweave) {
      this.bundlrClient = new Bundlr(
        config.bundlrNodeUrl,
        "arweave",
        config.arweaveJwkKey
      );
    }
  }

  async safelySaveDataPackages(dataPackages: CachedDataPackage[]) {
    try {
      await this.saveDataPackages(dataPackages);
    } catch (e) {
      this.logger.error("Error occured while saving to Bundlr", e.stack);
    }
  }

  async saveDataPackages(dataPackages: CachedDataPackage[]) {
    for (const dataPackage of dataPackages) {
      await this.saveOneDataPackage(dataPackage);
    }
  }

  async saveOneDataPackage(dataPackage: CachedDataPackage) {
    const timestampSeconds = Math.round(
      dataPackage.timestampMilliseconds / 1000
    );
    const tagsObj = {
      type: REDSTONE_TYPE_TAG_VALUE,
      timestamp: timestampSeconds.toString(),
      dataServiceId: dataPackage.dataServiceId,
      signerAddress: dataPackage.signerAddress,
      dataFeedId: dataPackage.dataFeedId || "",
    };
    const tagsArray = Object.entries(tagsObj).map(([name, value]) => ({
      name,
      value,
    }));

    const dataToSave = this.prepareDataToSave(dataPackage);

    const tx = await this.bundlrClient.createTransaction(dataToSave, {
      tags: tagsArray,
    });

    await tx.sign();
    this.logger.log(`Tx signed: ${tx.id}`);

    await tx.upload();
    this.logger.log(`Tx uploaded: ${tx.id}`);
  }

  // TOOD: maybe use gzip compression in future
  prepareDataToSave(dataPackage: CachedDataPackage) {
    return JSON.stringify(dataPackage);
  }
}
