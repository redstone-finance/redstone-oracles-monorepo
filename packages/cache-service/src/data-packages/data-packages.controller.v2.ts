import {
  Controller,
  Get,
  Header,
  Param,
  ServiceUnavailableException,
} from "@nestjs/common";
import config from "../config";
import { BaseDataPackagesController } from "./base-data-packages.controller";
import { DataPackagesResponse } from "./data-packages.interface";
import { DataPackagesService } from "./data-packages.service";

@Controller("v2/data-packages")
export class DataPackagesControllerV2 extends BaseDataPackagesController {
  protected readonly allowExternalSigners = true;

  @Get("latest/:DATA_SERVICE_ID/hide-metadata")
  @Header("Cache-Control", "max-age=5")
  async getAllLatestWithNoMetadata(
    @Param("DATA_SERVICE_ID") dataServiceId: string
  ): Promise<DataPackagesResponse> {
    await BaseDataPackagesController.validateDataServiceId(dataServiceId);
    return await this.dataPackagesService.getLatestDataPackagesWithSameTimestampWithCache(
      dataServiceId,
      true,
      this.allowExternalSigners
    );
  }

  @Get("historical/:DATA_SERVICE_ID/:TIMESTAMP/hide-metadata")
  @Header("Cache-Control", "max-age=5")
  async getDataPackagesByTimestampWithNoMetadata(
    @Param("DATA_SERVICE_ID") dataServiceId: string,
    @Param("TIMESTAMP") timestamp: string
  ): Promise<DataPackagesResponse> {
    if (!config.enableHistoricalDataServing) {
      throw new ServiceUnavailableException(
        `historical/* routes are not enabled in this cache-service configuration`
      );
    }
    await DataPackagesControllerV2.validateDataServiceId(dataServiceId);

    return await DataPackagesService.getDataPackagesByTimestamp(
      dataServiceId,
      Number(timestamp),
      true,
      this.allowExternalSigners
    );
  }
}
