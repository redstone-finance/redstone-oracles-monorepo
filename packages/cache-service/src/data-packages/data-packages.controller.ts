import {
  Body,
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Post,
  Query,
} from "@nestjs/common";
import { DataPackagesRequestParams } from "redstone-sdk";
import config from "../config";
import { ReceivedDataPackage } from "./data-packages.interface";
import { DataPackagesService } from "./data-packages.service";
import { CachedDataPackage } from "./data-packages.model";
import { BundlrService } from "../bundlr/bundlr.service";

export interface BulkPostRequestBody {
  requestSignature: string;
  dataPackages: ReceivedDataPackage[];
}

export interface GetLatestDataPackagesQuery {
  "data-service-id": string;
  "unique-signers-count": number;
  "data-feeds": string;
}

export interface DataPackagesResponse {
  [dataFeedId: string]: CachedDataPackage[];
}

@Controller("data-packages")
export class DataPackagesController {
  private bundlrService = new BundlrService();

  constructor(private dataPackagesService: DataPackagesService) {}

  @Get("latest")
  async getLatest(
    @Query() query: GetLatestDataPackagesQuery
  ): Promise<DataPackagesResponse> {
    // TODO: implement request validation
    const requestParams: DataPackagesRequestParams = {
      dataServiceId: query["data-service-id"],
      uniqueSignersCount: query["unique-signers-count"],
      dataFeeds: (query["data-feeds"] ?? "").split(","),
    };

    return await this.dataPackagesService.getDataPackages(requestParams);
  }

  @Post("bulk")
  async addBulk(@Body() body: BulkPostRequestBody) {
    if (!config.enableDirectPostingRoutes) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          error: "Data package posting routes are disabled",
        },
        HttpStatus.BAD_REQUEST
      );
    }

    // TODO: implement request validation
    // TODO: implement request size limit

    const signerAddress = await this.dataPackagesService.verifyRequester(body);

    const dataPackagesToSave =
      await this.dataPackagesService.prepareReceivedDataPackagesForBulkSaving(
        body.dataPackages,
        signerAddress
      );

    await this.dataPackagesService.saveManyDataPackagesInDB(dataPackagesToSave);

    if (config.enableArchivingOnArweave) {
      await this.bundlrService.safelySaveDataPackages(dataPackagesToSave);
    }
  }
}
