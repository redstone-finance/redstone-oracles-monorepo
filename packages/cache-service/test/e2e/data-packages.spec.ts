import "../common/set-test-envs";
import { signByMockSigner } from "../common/test-utils";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";
import * as request from "supertest";
import { AppModule } from "../../src/app.module";
import {
  mockDataPackages,
  mockOracleRegistryState,
  mockSigner,
} from "../common/mock-values";
import { connectToTestDB, dropTestDatabase } from "../common/test-db";
import { DataPackage } from "../../src/data-packages/data-packages.model";
import { BundlrService } from "../../src/bundlr/bundlr.service";

jest.mock("redstone-sdk", () => ({
  __esModule: true,
  ...jest.requireActual("redstone-sdk"),
  getOracleRegistryState: jest.fn(() => mockOracleRegistryState),
}));
jest.mock("../../src/bundlr/bundlr.service");

const expectedDataPackages = mockDataPackages.map((dataPackage) => ({
  ...dataPackage,
  signerAddress: mockSigner.address,
  dataServiceId: "mock-data-service-1",
}));

describe("Data packages (e2e)", () => {
  let app: INestApplication, httpServer: any;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
    httpServer = app.getHttpServer();

    (BundlrService.prototype.safelySaveDataPackages as any).mockClear();

    // Connect to mongoDB in memory
    await connectToTestDB();
  });

  afterEach(async () => await dropTestDatabase());

  it("/data-packages/bulk (POST)", async () => {
    const requestSignature = signByMockSigner(mockDataPackages);
    await request(httpServer)
      .post("/data-packages/bulk")
      .send({
        requestSignature,
        dataPackages: mockDataPackages,
      })
      .expect(201);

    const dataPackagesInDB = await DataPackage.find().sort({
      dataFeedId: 1,
    });
    const dataPackagesInDBCleaned = dataPackagesInDB.map((dp) => {
      const { _id, __v, ...rest } = dp.toJSON() as any;
      return rest;
    });
    expect(dataPackagesInDBCleaned).toEqual(expectedDataPackages);
  });

  it("/data-packages/bulk (POST) - should post data using Bundlr", async () => {
    const requestSignature = signByMockSigner(mockDataPackages);
    await request(httpServer)
      .post("/data-packages/bulk")
      .send({
        requestSignature,
        dataPackages: mockDataPackages,
      })
      .expect(201);

    // Should have been saved in Arweave
    expect(
      BundlrService.prototype.safelySaveDataPackages
    ).toHaveBeenCalledTimes(1);
    expect(BundlrService.prototype.safelySaveDataPackages).toHaveBeenCalledWith(
      expectedDataPackages
    );
  });

  it("/data-packages/bulk (POST) - should fail for invalid signature", async () => {
    const requestSignature = signByMockSigner(mockDataPackages);
    const newDataPackages = [...mockDataPackages];
    newDataPackages[0].dataPoints[0].value = 43;
    await request(httpServer)
      .post("/data-packages/bulk")
      .send({
        requestSignature,
        mockDataPackages,
      })
      .expect(500);

    expect(await DataPackage.find()).toEqual([]);
  });

  it("/data-packages/latest (GET)", async () => {
    const dataPackagesToInsert = [];
    for (const dataServiceId of ["service-1", "service-2", "service-3"]) {
      for (const dataFeedId of [undefined, "ETH", "AAVE", "BTC"]) {
        for (const signerAddress of ["0x1", "0x2", "0x3", "0x4", "0x5"]) {
          dataPackagesToInsert.push({
            ...mockDataPackages[0],
            dataFeedId,
            dataServiceId,
            signerAddress,
          });
        }
      }
    }
    await DataPackage.insertMany(dataPackagesToInsert);

    const testResponse = await request(httpServer)
      .get("/data-packages/latest")
      .query({
        "data-service-id": "service-2",
        "unique-signers-count": 4,
        "data-feeds": "ETH,BTC",
      })
      .expect(200);

    for (const dataFeedId of ["BTC", "ETH"]) {
      expect(testResponse.body[dataFeedId].length).toBe(4);
      const signers = [];
      for (const dataPackage of testResponse.body[dataFeedId]) {
        expect(dataPackage).toHaveProperty("dataFeedId", dataFeedId);
        expect(dataPackage).toHaveProperty("sources", null);
        expect(dataPackage).toHaveProperty("signature", "mock-signature");
        signers.push(dataPackage.signerAddress);
      }
      expect(signers.length).toBe(4);
      expect(new Set(signers).size).toBe(4);
    }
  });
});
