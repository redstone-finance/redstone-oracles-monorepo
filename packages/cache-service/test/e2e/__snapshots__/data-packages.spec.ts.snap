// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data packages (e2e) /data-packages/historical/mock-data-service-1 (GET): historical-data 1`] = `
[
  {
    "dataFeedId": "___ALL_FEEDS___",
    "dataPoints": [
      {
        "dataFeedId": "XD",
        "metadata": {
          "sources": {
            "exchange_1": "1.19",
            "exchange_2": "1.20",
            "exchange_3": "1.21",
          },
          "type": "HEX_BIG_INT",
        },
        "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcnDgA=",
      },
      {
        "dataFeedId": "mock-data-feed-id-1",
        "value": 42,
      },
      {
        "dataFeedId": "mock-data-feed-id-2",
        "value": 123,
      },
    ],
    "dataServiceId": "mock-data-service-1",
    "isSignatureValid": true,
    "signature": "PJkkAPFdAh24XW5IyiFbdL2Sjhsl7S7kLSGoiCACZZtGNjkeHXiFed7Zq4KP1pMVAanBp5B+qmtKb9Wht5RANxw=",
    "signerAddress": "0x2",
    "timestampMilliseconds": 1654353399000,
  },
  {
    "dataFeedId": "___ALL_FEEDS___",
    "dataPoints": [
      {
        "dataFeedId": "XD",
        "metadata": {
          "sources": {
            "exchange_1": "1.19",
            "exchange_2": "1.20",
            "exchange_3": "1.21",
          },
          "type": "HEX_BIG_INT",
        },
        "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcnDgA=",
      },
      {
        "dataFeedId": "mock-data-feed-id-1",
        "value": 42,
      },
      {
        "dataFeedId": "mock-data-feed-id-2",
        "value": 123,
      },
    ],
    "dataServiceId": "mock-data-service-1",
    "isSignatureValid": true,
    "signature": "PJkkAPFdAh24XW5IyiFbdL2Sjhsl7S7kLSGoiCACZZtGNjkeHXiFed7Zq4KP1pMVAanBp5B+qmtKb9Wht5RANxw=",
    "signerAddress": "0x3",
    "timestampMilliseconds": 1654353399000,
  },
  {
    "dataFeedId": "___ALL_FEEDS___",
    "dataPoints": [
      {
        "dataFeedId": "XD",
        "metadata": {
          "sources": {
            "exchange_1": "1.19",
            "exchange_2": "1.20",
            "exchange_3": "1.21",
          },
          "type": "HEX_BIG_INT",
        },
        "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcnDgA=",
      },
      {
        "dataFeedId": "mock-data-feed-id-1",
        "value": 42,
      },
      {
        "dataFeedId": "mock-data-feed-id-2",
        "value": 123,
      },
    ],
    "dataServiceId": "mock-data-service-1",
    "isSignatureValid": true,
    "signature": "PJkkAPFdAh24XW5IyiFbdL2Sjhsl7S7kLSGoiCACZZtGNjkeHXiFed7Zq4KP1pMVAanBp5B+qmtKb9Wht5RANxw=",
    "signerAddress": "0x4",
    "timestampMilliseconds": 1654353399000,
  },
  {
    "dataFeedId": "___ALL_FEEDS___",
    "dataPoints": [
      {
        "dataFeedId": "XD",
        "metadata": {
          "sources": {
            "exchange_1": "1.19",
            "exchange_2": "1.20",
            "exchange_3": "1.21",
          },
          "type": "HEX_BIG_INT",
        },
        "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcnDgA=",
      },
      {
        "dataFeedId": "mock-data-feed-id-1",
        "value": 42,
      },
      {
        "dataFeedId": "mock-data-feed-id-2",
        "value": 123,
      },
    ],
    "dataServiceId": "mock-data-service-1",
    "isSignatureValid": true,
    "signature": "PJkkAPFdAh24XW5IyiFbdL2Sjhsl7S7kLSGoiCACZZtGNjkeHXiFed7Zq4KP1pMVAanBp5B+qmtKb9Wht5RANxw=",
    "signerAddress": "0x5",
    "timestampMilliseconds": 1654353399000,
  },
  {
    "dataFeedId": "___ALL_FEEDS___",
    "dataPoints": [
      {
        "dataFeedId": "XD",
        "metadata": {
          "sources": {
            "exchange_1": "1.19",
            "exchange_2": "1.20",
            "exchange_3": "1.21",
          },
          "type": "HEX_BIG_INT",
        },
        "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcnDgA=",
      },
      {
        "dataFeedId": "mock-data-feed-id-1",
        "value": 42,
      },
      {
        "dataFeedId": "mock-data-feed-id-2",
        "value": 123,
      },
    ],
    "dataServiceId": "mock-data-service-1",
    "isSignatureValid": true,
    "signature": "PJkkAPFdAh24XW5IyiFbdL2Sjhsl7S7kLSGoiCACZZtGNjkeHXiFed7Zq4KP1pMVAanBp5B+qmtKb9Wht5RANxw=",
    "signerAddress": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "timestampMilliseconds": 1654353399000,
  },
]
`;
